File: src/46_permutations/permutations.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

class Solution {
   public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        backtrack(nums, 0, res);
        return res;
    }
    void backtrack(vector<int>& nums, int i, vector<vector<int>>& res) {
        if (i == nums.size()) {
            res.push_back(nums);
            return;
        }
        for (int j = i; j < nums.size(); ++j) {
            swap(nums[i], nums[j]);
            backtrack(nums, i + 1, res);
            swap(nums[i], nums[j]);
        }
    }
    // Iterative version
    vector<vector<int>> permuteIterative(vector<int> nums) {
        vector<vector<int>> res = {{}};
        for (int num : nums) {
            vector<vector<int>> nextPerms;
            for (auto& p : res) {
                for (int i = 0; i <= p.size(); ++i) {
                    vector<int> pCopy = p;
                    pCopy.insert(pCopy.begin() + i, num);
                    nextPerms.push_back(pCopy);
                }
            }
            res = nextPerms;
        }
        return res;
    }
};

string formatResult(vector<vector<int>> result) {
    vector<string> sorted;
    for (auto& perm : result) {
        sort(perm.begin(), perm.end());
        string s = "[";
        for (size_t i = 0; i < perm.size(); i++) {
            s += to_string(perm[i]);
            if (i < perm.size() - 1) s += ",";
        }
        s += "]";
        sorted.push_back(s);
    }
    sort(sorted.begin(), sorted.end());
    string res = "[";
    for (size_t i = 0; i < sorted.size(); i++) {
        res += sorted[i];
        if (i < sorted.size() - 1) res += ",";
    }
    res += "]";
    return res;
}

int main() {
    Solution sol;
    // Test case 1
    vector<int> nums1 = {1, 2, 3};
    vector<vector<int>> e1 = {{1, 2, 3}, {1, 3, 2}, {2, 1, 3}, {2, 3, 1}, {3, 1, 2}, {3, 2, 1}};
    cout << formatResult(sol.permute(nums1)) << " | " << formatResult(e1) << endl;
    // Test case 2
    vector<int> nums2 = {0, 1};
    vector<vector<int>> e2 = {{0, 1}, {1, 0}};
    cout << formatResult(sol.permute(nums2)) << " | " << formatResult(e2) << endl;
    // Test case 3
    vector<int> nums3 = {1};
    vector<vector<int>> e3 = {{1}};
    cout << formatResult(sol.permute(nums3)) << " | " << formatResult(e3) << endl;
    // Test case 4
    vector<int> nums4 = {2, 1};
    vector<vector<int>> e4 = {{2, 1}, {1, 2}};
    cout << formatResult(sol.permute(nums4)) << " | " << formatResult(e4) << endl;
    // Test case 5
    vector<int> nums5 = {1, 2, 3, 4};
    cout << sol.permute(nums5).size() << " | " << 24 << endl;
    return 0;
}

File: src/46_permutations/Permutations.java
import java.util.*;

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        backtrack(nums, 0, res);
        return res;
    }

    private void backtrack(int[] nums, int i, List<List<Integer>> res) {
        if (i == nums.length) {
            List<Integer> perm = new ArrayList<>();
            for (int n : nums)
                perm.add(n);
            res.add(new ArrayList<>(perm));
            return;
        }
        for (int j = i; j < nums.length; ++j) {
            swap(nums, i, j);
            backtrack(nums, i + 1, res);
            swap(nums, i, j);
        }
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    public List<List<Integer>> permuteIterative(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        res.add(new ArrayList<>());
        for (int num : nums) {
            List<List<Integer>> nextPerms = new ArrayList<>();
            for (List<Integer> p : res) {
                for (int i = 0; i <= p.size(); ++i) {
                    List<Integer> pCopy = new ArrayList<>(p);
                    pCopy.add(i, num);
                    nextPerms.add(pCopy);
                }
            }
            res = nextPerms;
        }
        return res;
    }
}

public class Permutations {
    private static String formatResult(List<List<Integer>> result) {
        List<String> sorted = new ArrayList<>();
        for (List<Integer> perm : result) {
            Collections.sort(perm);
            sorted.add(perm.toString());
        }
        Collections.sort(sorted);
        return sorted.toString();
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        // Test case 1
        int[] nums1 = { 1, 2, 3 };
        List<List<Integer>> e1 = Arrays.asList(
                Arrays.asList(1, 2, 3), Arrays.asList(1, 3, 2), Arrays.asList(2, 1, 3), Arrays.asList(2, 3, 1),
                Arrays.asList(3, 1, 2), Arrays.asList(3, 2, 1));
        System.out.println(formatResult(sol.permute(nums1)) + " | " + formatResult(e1));
        // Test case 2
        int[] nums2 = { 0, 1 };
        List<List<Integer>> e2 = Arrays.asList(Arrays.asList(0, 1), Arrays.asList(1, 0));
        System.out.println(formatResult(sol.permute(nums2)) + " | " + formatResult(e2));
        // Test case 3
        int[] nums3 = { 1 };
        List<List<Integer>> e3 = Arrays.asList(Arrays.asList(1));
        System.out.println(formatResult(sol.permute(nums3)) + " | " + formatResult(e3));
        // Test case 4
        int[] nums4 = { 2, 1 };
        List<List<Integer>> e4 = Arrays.asList(Arrays.asList(2, 1), Arrays.asList(1, 2));
        System.out.println(formatResult(sol.permute(nums4)) + " | " + formatResult(e4));
        // Test case 5
        int[] nums5 = { 1, 2, 3, 4 };
        System.out.println((sol.permute(nums5) != null ? sol.permute(nums5).size() : 0) + " | " + 24);
    }
}
